packages/xsnap/src/xsnap.c

```diff
@@ -0,0 +1,1350 @@
+#include "xsAll.h"
+#include "xsScript.h"
+#include "xsSnapshot.h"
+#include "xs.h"
+
+extern txScript* fxLoadScript(txMachine* the, txString path, txUnsigned flags);
+
+typedef struct sxAliasIDLink txAliasIDLink;
+typedef struct sxAliasIDList txAliasIDList;
+typedef struct sxJob txJob;
+typedef void (*txJobCallback)(txJob*);
+
+struct sxAliasIDLink {
+	txAliasIDLink* previous;
+	txAliasIDLink* next;
+	txInteger id;
+	txInteger flag;
+};
+
+struct sxAliasIDList {
+	txAliasIDLink* first;
+	txAliasIDLink* last;
+	txFlag* aliases;
+	txInteger errorCount;
+};
+
+struct sxJob {
+	txJob* next;
+	txMachine* the;
+	txNumber when;
+	txJobCallback callback;
+	txSlot function;
+	txSlot argument;
+	txNumber interval;
+};
+
+static void fxBuildAgent(xsMachine* the);
+static txInteger fxCheckAliases(txMachine* the);
+static void fxCheckAliasesError(txMachine* the, txAliasIDList* list, txFlag flag);
+static void fxCheckEnvironmentAliases(txMachine* the, txSlot* environment, txAliasIDList* list);
+static void fxCheckInstanceAliases(txMachine* the, txSlot* instance, txAliasIDList* list);
+static void fxFreezeBuiltIns(txMachine* the);
+static void fxPatchBuiltIns(txMachine* the);
+static void fxPrintUsage();
+
+static void fx_Array_prototype_meter(xsMachine* the);
+
+extern void fx_clearTimer(txMachine* the);
+static void fx_destroyTimer(void* data);
+static void fx_evalScript(xsMachine* the);
+static void fx_gc(xsMachine* the);
+static void fx_isPromiseJobQueueEmpty(xsMachine* the);
+static void fx_markTimer(txMachine* the, void* it, txMarkRoot markRoot);
+static void fx_print(xsMachine* the);
+static void fx_setImmediate(txMachine* the);
+static void fx_setInterval(txMachine* the);
+static void fx_setTimeout(txMachine* the);
+static void fx_setTimer(txMachine* the, txNumber interval, txBoolean repeat);
+static void fx_setTimerCallback(txJob* job);
+static void fx_sysCall(xsMachine *the);
+
+static void fxFulfillModuleFile(txMachine* the);
+static void fxRejectModuleFile(txMachine* the);
+static void fxRunModuleFile(txMachine* the, txString path);
+static void fxRunProgramFile(txMachine* the, txString path, txUnsigned flags);
+static void fxRunLoop(txMachine* the);
+
+static int fxReadNetString(FILE *inStream, char** dest, size_t* len);
+static char* fxReadNetStringError(int code);
+static int fxWriteNetString(FILE* outStream, char prefix, char* buf, size_t len);
+static char* fxWriteNetStringError(int code);
+
+// The order of the callbacks materially affects how they are introduced to
+// code that runs from a snapshot, so must be consistent in the face of
+// upgrade.
+#define mxSnapshotCallbackCount 10
+txCallback gxSnapshotCallbacks[mxSnapshotCallbackCount] = {
+	fx_clearTimer, // 0
+	fx_evalScript, // 1
+	fx_gc, // 2
+	fx_isPromiseJobQueueEmpty, // 3
+	fx_print, // 4
+	fx_sysCall, // 5
+	fx_setImmediate, // 6
+	fx_setInterval, // 7
+	fx_setTimeout, // 8
+	fx_Array_prototype_meter, // 9
+};
```

Comment:
Added markings here to make this more resilient in the face of maintenance. Like Protobuf field offsets, moving these around breaks snapshot compatibility. We cannot remove, but we can change the behavior of a thunk, albeit to throw an error.

Reply:
That list looks like a prime candidate for an automated test.

Even though we can, as you point out, replace these with error stubs, I'm pretty sure we don't need `fx_isPromiseJobQueueEmpty` and I wonder if we want to include it in this initial list.

`gc` is a bit of a footgun in that we need to be sure it's not exposed to deterministic code and it's not one of the usual suspects such as `setImmediate`. I guess SES will take care of it, but that seems worth testing. Though this will become evident in a PR coming very soon, I don't trust myself to remember even that long. I'd like someone else to weigh in. @warner ?

Reply:
There’s a reasonably good chance that we’ll invoke `gc` from C instead of JS. I’ll comment out all the bits we don’t need yet, including all the timers and print.